OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,454 @@
+                                                 .alpha = marker_info->color[3],
+                                             });
+  }
+
+  void OnCmdDebugMarkerEndEXT(VkCommandBuffer command_buffer) {
+    submission_tracker_.MarkDebugMarkerEnd(command_buffer);
+    if (dispatch_table_.IsDebugMarkerExtensionSupported(command_buffer)) {
+      dispatch_table_.CmdDebugMarkerEndEXT(command_buffer)(command_buffer);
+    }
+  }
+
+  // ----------------------------------------------------------------------------
+  // Layer enumeration functions
+  // ----------------------------------------------------------------------------
+
+  [[nodiscard]] VkResult OnEnumerateInstanceLayerProperties(uint32_t* property_count,
+                                                            VkLayerProperties* properties) {
+    // Vulkan spec dictates that we are only supposed to enumerate ourselve.
+    if (property_count != nullptr) {
+      *property_count = 1;
+    }
+    if (properties != nullptr) {
+      snprintf(properties->layerName, sizeof(properties->layerName), kLayerName);
+      snprintf(properties->description, sizeof(properties->description), kLayerDescription);
+      properties->implementationVersion = kLayerImplVersion;
+      properties->specVersion = kLayerSpecVersion;
+    }
+
+    return VK_SUCCESS;
+  }
+
+  [[nodiscard]] VkResult OnEnumerateInstanceExtensionProperties(
+      const char* layer_name, uint32_t* property_count, VkExtensionProperties* /*properties*/) {
+    // Inform the client that we have no extension properties if this layer
+    // specifically is being queried.
+    if (layer_name != nullptr && strcmp(layer_name, kLayerName) == 0) {
+      if (property_count != nullptr) {
+        *property_count = 0;
+      }
+      return VK_SUCCESS;
+    }
+
+    // Vulkan spec mandates returning this when this layer isn't being queried.
+    return VK_ERROR_LAYER_NOT_PRESENT;
+  }
+
+  [[nodiscard]] VkResult OnEnumerateDeviceExtensionProperties(VkPhysicalDevice physical_device,
+                                                              const char* layer_name,
+                                                              uint32_t* property_count,
+                                                              VkExtensionProperties* properties) {
+    // If our layer is queried exclusively, we just return our extensions
+    if (layer_name != nullptr && strcmp(layer_name, kLayerName) == 0) {
+      // If properties == nullptr, only the number of extensions are queried.
+      if (properties == nullptr) {
+        *property_count = kDeviceExtensions.size();
+        return VK_SUCCESS;
+      }
+      uint32_t num_extensions_to_copy = kDeviceExtensions.size();
+      // In the case that less extensions are queried then the layer uses, we copy on this number
+      // and return VK_INCOMPLETE, according to the specification.
+      if (*property_count < num_extensions_to_copy) {
```

Comment:
I'm confused here. So `property_count` is both an in and an out parameter?

Reply:
Yes, it does. For all those functions the pattern is usually the following:
```
vkFoo(&count, nullptr);
// Allocate an array "data" of size "count"
vkFoo(&count,  &data);
```
In that particular function, the count also acts as input. So if the requested count is smaller than the actual count, it only writes the smaller number of elements.